From 148424d61fc47e0cddd22633d5468c0fc634d8b5 Mon Sep 17 00:00:00 2001
From: "jianqin.zhang" <jianqin.zhang@amlogic.com>
Date: Mon, 3 Nov 2014 13:53:57 +0800
Subject: [PATCH 5008/5965] pd#93522:add dual display function for HDMI and
 CVBS

Conflicts:
	arch/arm/boot/dts/amlogic/meson8_k200a_1G.dtd
	arch/arm/boot/dts/amlogic/meson8_k200a_2G.dtd
	arch/arm/boot/dts/amlogic/meson8_k200b_1G_emmc_sdhc.dtd
	arch/arm/boot/dts/amlogic/meson8_k200b_1G_emmc_sdio.dtd
	arch/arm/boot/dts/amlogic/meson8_k200b_2G_emmc_sdhc.dtd
	arch/arm/boot/dts/amlogic/meson8_k200b_2G_emmc_sdio.dtd

Change-Id: I672ef118471753b3e19a30e6ac5d40d456dca2e3
---
 arch/arm/boot/dts/amlogic/meson8_k200a_1G.dtd |  24 +-
 arch/arm/boot/dts/amlogic/meson8_k200a_2G.dtd |  25 ++-
 .../dts/amlogic/meson8_k200b_1G_emmc_sdhc.dtd |  25 ++-
 .../dts/amlogic/meson8_k200b_1G_emmc_sdio.dtd |  24 +-
 .../dts/amlogic/meson8_k200b_2G_emmc_sdhc.dtd |  24 +-
 .../dts/amlogic/meson8_k200b_2G_emmc_sdio.dtd |  24 +-
 arch/arm/configs/meson8_defconfig             |   3 +
 drivers/amlogic/amports/Kconfig               |   6 +
 drivers/amlogic/amports/video.c               |   6 +-
 drivers/amlogic/amports/video2.c              | 144 ++++++++++--
 drivers/amlogic/display/vout2/nulldisp.c      |   7 +-
 drivers/amlogic/display/vout2/tvconf2.c       | 212 +++++++++++++++++-
 drivers/amlogic/display/vout2/tvmode.h        |  13 ++
 drivers/amlogic/display/vout2/tvoutc2.c       |  43 +++-
 drivers/amlogic/display/vout2/tvregs.h        |  55 ++++-
 drivers/amlogic/display/vout2/vout2_serve.c   |  26 ++-
 drivers/amlogic/display/vout2/vout_serve.h    |   2 +-
 drivers/amlogic/vfm/vfm.c                     |   2 +-
 18 files changed, 606 insertions(+), 59 deletions(-)

diff --git a/arch/arm/boot/dts/amlogic/meson8_k200a_1G.dtd b/arch/arm/boot/dts/amlogic/meson8_k200a_1G.dtd
index 99bf0e1e8bb7..27f745060b4c 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k200a_1G.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k200a_1G.dtd
@@ -241,7 +241,22 @@ void root_func(){
 //	
 /// ***************************************************************************************
 ///	-	DISP&MM-VDIN
-//$$ MODULE = "DISP&MM-VDIN"	
+//$$ MODULE = "DISP&MM-VDIN"
+//$$ DEVICE="vdin0"
+//$$ L2 PROP_STR = "status"
+//$$ L2 PROP_U32 2 = "reg"
+//$$ L2 PROP_U32 = "irq"
+//$$ L2 PROP_U32 = "vdin_id"
+	vdin0{
+		compatible = "amlogic,vdin";
+		dev_name = "vdin0";
+		status = "ok";
+		reserve-memory = <0x01000000>;
+		reserve-iomap = "true";
+		irq = <115>;
+		vdin_id = <0>;
+	};
+
 //$$ DEVICE="vdin1"
 //$$ L2 PROP_STR = "status"
 //$$ L2 PROP_U32 2 = "reg"
@@ -296,6 +311,13 @@ void root_func(){
 		status = "okay";
 	};
 
+//$$ DEVICE = "mesonvout2"
+//$$ L2 PROP_STR = "status"
+	mesonvout2{
+		compatible = "amlogic,mesonvout2";
+		dev_name = "mesonvout2";
+		status = "okay";
+	};
 /// ***************************************************************************************
 ///	-	DISP&MM-A/V Amvideocap
 //$$ MODULE = "DISP&MM-Amvideocap"
diff --git a/arch/arm/boot/dts/amlogic/meson8_k200a_2G.dtd b/arch/arm/boot/dts/amlogic/meson8_k200a_2G.dtd
index de8907061038..3681f0058861 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k200a_2G.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k200a_2G.dtd
@@ -220,7 +220,22 @@ void root_func(){
 //	
 /// ***************************************************************************************
 ///	-	DISP&MM-VDIN
-//$$ MODULE = "DISP&MM-VDIN"	
+//$$ MODULE = "DISP&MM-VDIN"
+//$$ DEVICE="vdin0"
+//$$ L2 PROP_STR = "status"
+//$$ L2 PROP_U32 2 = "reg"
+//$$ L2 PROP_U32 = "irq"
+//$$ L2 PROP_U32 = "vdin_id"
+	vdin0{
+		compatible = "amlogic,vdin";
+		dev_name = "vdin0";
+		status = "ok";
+		reserve-memory = <0x01000000>;
+		reserve-iomap = "true";
+		irq = <115>;
+		vdin_id = <0>;
+	};
+
 //$$ DEVICE="vdin1"
 //$$ L2 PROP_STR = "status"
 //$$ L2 PROP_U32 2 = "reg"
@@ -275,6 +290,14 @@ void root_func(){
 		status = "okay";
 	};
 
+//$$ DEVICE = "mesonvout2"
+//$$ L2 PROP_STR = "status"
+	mesonvout2{
+		compatible = "amlogic,mesonvout2";
+		dev_name = "mesonvout2";
+		status = "okay";
+	};
+
 /// ***************************************************************************************
 ///	-	DISP&MM-A/V Amvideocap
 //$$ MODULE = "DISP&MM-Amvideocap"
diff --git a/arch/arm/boot/dts/amlogic/meson8_k200b_1G_emmc_sdhc.dtd b/arch/arm/boot/dts/amlogic/meson8_k200b_1G_emmc_sdhc.dtd
index cbfbee34e78e..9e6fc7dccf6e 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k200b_1G_emmc_sdhc.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k200b_1G_emmc_sdhc.dtd
@@ -254,7 +254,22 @@ void root_func(){
 //	
 /// ***************************************************************************************
 ///	-	DISP&MM-VDIN
-//$$ MODULE = "DISP&MM-VDIN"	
+//$$ MODULE = "DISP&MM-VDIN"
+//$$ DEVICE="vdin0"
+//$$ L2 PROP_STR = "status"
+//$$ L2 PROP_U32 2 = "reg"
+//$$ L2 PROP_U32 = "irq"
+//$$ L2 PROP_U32 = "vdin_id"
+	vdin0{
+		compatible = "amlogic,vdin";
+		dev_name = "vdin0";
+		status = "ok";
+		reserve-memory = <0x01000000>;
+		reserve-iomap = "true";
+		irq = <115>;
+		vdin_id = <0>;
+	};
+
 //$$ DEVICE="vdin1"
 //$$ L2 PROP_STR = "status"
 //$$ L2 PROP_U32 2 = "reg"
@@ -309,6 +324,14 @@ void root_func(){
 		status = "okay";
 	};
 
+//$$ DEVICE = "mesonvout2"
+//$$ L2 PROP_STR = "status"
+	mesonvout2{
+		compatible = "amlogic,mesonvout2";
+		dev_name = "mesonvout2";
+		status = "okay";
+	};
+
 /// ***************************************************************************************
 ///	-	DISP&MM-A/V Amvideocap
 //$$ MODULE = "DISP&MM-Amvideocap"
diff --git a/arch/arm/boot/dts/amlogic/meson8_k200b_1G_emmc_sdio.dtd b/arch/arm/boot/dts/amlogic/meson8_k200b_1G_emmc_sdio.dtd
index d16f6bc28eca..48fade445067 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k200b_1G_emmc_sdio.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k200b_1G_emmc_sdio.dtd
@@ -243,7 +243,22 @@ void root_func(){
 //	
 /// ***************************************************************************************
 ///	-	DISP&MM-VDIN
-//$$ MODULE = "DISP&MM-VDIN"	
+//$$ MODULE = "DISP&MM-VDIN"
+//$$ DEVICE="vdin0"
+//$$ L2 PROP_STR = "status"
+//$$ L2 PROP_U32 2 = "reg"
+//$$ L2 PROP_U32 = "irq"
+//$$ L2 PROP_U32 = "vdin_id"
+	vdin0{
+		compatible = "amlogic,vdin";
+		dev_name = "vdin0";
+		status = "ok";
+		reserve-memory = <0x01000000>;
+		reserve-iomap = "true";
+		irq = <115>;
+		vdin_id = <0>;
+	};
+
 //$$ DEVICE="vdin1"
 //$$ L2 PROP_STR = "status"
 //$$ L2 PROP_U32 2 = "reg"
@@ -298,6 +313,13 @@ void root_func(){
 		status = "okay";
 	};
 
+//$$ DEVICE = "mesonvout2"
+//$$ L2 PROP_STR = "status"
+	mesonvout2{
+		compatible = "amlogic,mesonvout2";
+		dev_name = "mesonvout2";
+		status = "okay";
+	};
 /// ***************************************************************************************
 ///	-	DISP&MM-A/V Amvideocap
 //$$ MODULE = "DISP&MM-Amvideocap"
diff --git a/arch/arm/boot/dts/amlogic/meson8_k200b_2G_emmc_sdhc.dtd b/arch/arm/boot/dts/amlogic/meson8_k200b_2G_emmc_sdhc.dtd
index 1595bf319f06..ba7f305e83fa 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k200b_2G_emmc_sdhc.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k200b_2G_emmc_sdhc.dtd
@@ -220,7 +220,22 @@ void root_func(){
 //	
 /// ***************************************************************************************
 ///	-	DISP&MM-VDIN
-//$$ MODULE = "DISP&MM-VDIN"	
+//$$ MODULE = "DISP&MM-VDIN"
+//$$ DEVICE="vdin0"
+//$$ L2 PROP_STR = "status"
+//$$ L2 PROP_U32 2 = "reg"
+//$$ L2 PROP_U32 = "irq"
+//$$ L2 PROP_U32 = "vdin_id"
+	vdin0{
+		compatible = "amlogic,vdin";
+		dev_name = "vdin0";
+		status = "ok";
+		reserve-memory = <0x01000000>;
+		reserve-iomap = "true";
+		irq = <115>;
+		vdin_id = <0>;
+	};
+
 //$$ DEVICE="vdin1"
 //$$ L2 PROP_STR = "status"
 //$$ L2 PROP_U32 2 = "reg"
@@ -275,6 +290,13 @@ void root_func(){
 		status = "okay";
 	};
 
+//$$ DEVICE = "mesonvout2"
+//$$ L2 PROP_STR = "status"
+	mesonvout2{
+		compatible = "amlogic,mesonvout2";
+		dev_name = "mesonvout2";
+		status = "okay";
+	};
 /// ***************************************************************************************
 ///	-	DISP&MM-A/V Amvideocap
 //$$ MODULE = "DISP&MM-Amvideocap"
diff --git a/arch/arm/boot/dts/amlogic/meson8_k200b_2G_emmc_sdio.dtd b/arch/arm/boot/dts/amlogic/meson8_k200b_2G_emmc_sdio.dtd
index bff6e455b98d..1affce2e20c4 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k200b_2G_emmc_sdio.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k200b_2G_emmc_sdio.dtd
@@ -220,7 +220,22 @@ void root_func(){
 //	
 /// ***************************************************************************************
 ///	-	DISP&MM-VDIN
-//$$ MODULE = "DISP&MM-VDIN"	
+//$$ MODULE = "DISP&MM-VDIN"
+//$$ DEVICE="vdin0"
+//$$ L2 PROP_STR = "status"
+//$$ L2 PROP_U32 2 = "reg"
+//$$ L2 PROP_U32 = "irq"
+//$$ L2 PROP_U32 = "vdin_id"
+	vdin0{
+		compatible = "amlogic,vdin";
+		dev_name = "vdin0";
+		status = "ok";
+		reserve-memory = <0x01000000>;
+		reserve-iomap = "true";
+		irq = <115>;
+		vdin_id = <0>;
+	};
+
 //$$ DEVICE="vdin1"
 //$$ L2 PROP_STR = "status"
 //$$ L2 PROP_U32 2 = "reg"
@@ -275,6 +290,13 @@ void root_func(){
 		status = "okay";
 	};
 
+//$$ DEVICE = "mesonvout2"
+//$$ L2 PROP_STR = "status"
+	mesonvout2{
+		compatible = "amlogic,mesonvout2";
+		dev_name = "mesonvout2";
+		status = "okay";
+	};
 /// ***************************************************************************************
 ///	-	DISP&MM-A/V Amvideocap
 //$$ MODULE = "DISP&MM-Amvideocap"
diff --git a/arch/arm/configs/meson8_defconfig b/arch/arm/configs/meson8_defconfig
index 41c6957e832f..906671e64312 100755
--- a/arch/arm/configs/meson8_defconfig
+++ b/arch/arm/configs/meson8_defconfig
@@ -239,6 +239,9 @@ CONFIG_AM_VIDEO=y
 CONFIG_AM_ENCODER=y
 CONFIG_AM_JPEG_ENCODER=y
 CONFIG_AM_TV_OUTPUT2=y
+CONFIG_AM_VIDEO2=y
+CONFIG_TVIN_VIUIN=y
+CONFIG_TVIN_VDIN=y
 CONFIG_FB_AM=y
 CONFIG_FB_OSD2_CURSOR=y
 CONFIG_FB_OSD2_DEFAULT_WIDTH=32
diff --git a/drivers/amlogic/amports/Kconfig b/drivers/amlogic/amports/Kconfig
index e06a0a161296..591f176fb117 100755
--- a/drivers/amlogic/amports/Kconfig
+++ b/drivers/amlogic/amports/Kconfig
@@ -14,6 +14,12 @@ config  VSYNC_RDMA
 	default n
 	help
 	  Select to enable VSYNC CBUS RDMA.
+
+config TVIN_VIUIN
+	bool "Amlogic TVIN VIUIN device"
+	default n
+	help
+	  VIUIN device driver.
 	
 config AM_VIDEO
 	bool "Video management"
diff --git a/drivers/amlogic/amports/video.c b/drivers/amlogic/amports/video.c
index 2ad543b4b2f5..9698bcb0fa22 100755
--- a/drivers/amlogic/amports/video.c
+++ b/drivers/amlogic/amports/video.c
@@ -2339,6 +2339,7 @@ static irqreturn_t vsync_isr(int irq, void *dev_id)
     if(vdin_ops){
 	arg.cmd = VDIN_CMD_ISR;
 	vdin_ops->tvin_vdin_func(1,&arg);
+	vdin_ops->tvin_vdin_func(0,&arg);
     }
 #endif
     vout_type = detect_vout_type();
@@ -3177,6 +3178,9 @@ static void video_vf_light_unreg_provider(void)
 
 static int video_receiver_event_fun(int type, void* data, void* private_data)
 {
+#ifdef CONFIG_AM_VIDEO2
+    char* provider_name;
+#endif
     if(type == VFRAME_EVENT_PROVIDER_UNREG){
         video_vf_unreg_provider();
 #ifdef CONFIG_AM_VIDEO2
@@ -3192,7 +3196,7 @@ static int video_receiver_event_fun(int type, void* data, void* private_data)
     else if(type == VFRAME_EVENT_PROVIDER_REG){
         enable_video_discontinue_report = 1;
 #ifdef CONFIG_AM_VIDEO2
-        char* provider_name = (char*)data;
+        provider_name = (char*)data;
         if(strncmp(provider_name, "decoder", 7)==0
             || strncmp(provider_name, "ppmgr", 5)==0
             || strncmp(provider_name, "deinterlace", 11)==0
diff --git a/drivers/amlogic/amports/video2.c b/drivers/amlogic/amports/video2.c
index 35ad9097ad60..36c763569811 100755
--- a/drivers/amlogic/amports/video2.c
+++ b/drivers/amlogic/amports/video2.c
@@ -31,7 +31,7 @@
 #include <linux/string.h>
 #include <linux/io.h>
 #include <linux/mm.h>
-#include <linux/major.h>
+#include <linux/amlogic/major.h>
 #include <linux/err.h>
 #include <linux/mutex.h>
 #include <linux/platform_device.h>
@@ -168,7 +168,7 @@ static int debug_flag = DEBUG_FLAG_BLACKOUT;
 #define MODULE_NAME "amvideo2"
 #define DEVICE_NAME "amvideo2"
 
-#ifdef CONFIG_AML_VSYNC_FIQ_ENABLE
+#if 0//def CONFIG_AML_VSYNC_FIQ_ENABLE
 #define FIQ_VSYNC
 #else
 #undef FIQ_VSYNC
@@ -356,7 +356,7 @@ static u32 disable_video = VIDEO_DISABLE_NONE;
 static u32 frame_repeat_count = 0;
 #endif
 
-static u32 clone = 1;
+static u32 clone = 0;
 static u32 stream_play_enable = 0;
 static u32 clone_vpts_remainder;
 static int clone_frame_rate_delay = 0;
@@ -1353,7 +1353,7 @@ static irqreturn_t vsync_isr(int irq, void *dev_id)
             }
             else if(clone_vpts_remainder < vsync_pts_inc){
                 vf = video_vf_get();
-#ifdef CONFIG_TVIN_VIUIN
+#if 0//def CONFIG_TVIN_VIUIN
 				if(vf->width > 1280){
 					if(clone_frame_scale_width != 0){
 						vdin0_set_hscale(
@@ -2446,12 +2446,81 @@ static ssize_t video_disable_store(struct class *cla, struct class_attribute *at
     return count;
 }
 
+static void output_axis_adjust(int src_w, int src_h, int* dst_w, int* dst_h, int angle)
+{
+    int w = 0, h = 0,disp_w = 0, disp_h =0;
+    disp_w = *dst_w;
+    disp_h = *dst_h;
+    if (angle %180 !=0) {
+        h = min((int)src_w, disp_h);
+        w = src_h * h / src_w;
+        if(w > disp_w ){
+            h = (h * disp_w)/w ;
+            w = disp_w;
+        }
+    }else{
+        if ((src_w < disp_w) && (src_h < disp_h)) {
+            w = src_w;
+            h = src_h;
+        } else if ((src_w * disp_h) > (disp_w * src_h)) {
+            w = disp_w;
+            h = disp_w * src_h / src_w;
+        } else {
+            h = disp_h;
+            w = disp_h * src_w / src_h;
+        }
+    }
+    *dst_w = w;
+    *dst_h = h;
+}
+
+static tvin_scan_mode_t vmode2scan_mode(vmode_t mode)
+{
+    tvin_scan_mode_t scan_mode = TVIN_SCAN_MODE_NULL;//1: progressive 2:interlaced
+    switch (mode) {
+        case VMODE_480I:
+        case VMODE_480CVBS:
+        case VMODE_576I:
+        case VMODE_576CVBS:
+        case VMODE_1080I:
+        case VMODE_1080I_50HZ:
+            scan_mode = TVIN_SCAN_MODE_INTERLACED;
+            break;
+        case VMODE_480P:
+        case VMODE_576P:
+        case VMODE_720P:
+        case VMODE_1080P:
+        case VMODE_720P_50HZ:
+        case VMODE_1080P_50HZ:
+        case VMODE_1080P_24HZ:
+        case VMODE_4K2K_30HZ:
+        case VMODE_4K2K_25HZ:
+        case VMODE_4K2K_24HZ:
+        case VMODE_4K2K_SMPTE:
+        case VMODE_VGA:
+        case VMODE_SVGA:
+        case VMODE_XGA:
+        case VMODE_SXGA:
+        case VMODE_LCD:
+        case VMODE_LVDS_1080P:
+        case VMODE_LVDS_1080P_50HZ:
+        case VMODE_LVDS_768P:
+            scan_mode = TVIN_SCAN_MODE_PROGRESSIVE;
+            break;
+        default:
+            printk("unknown mode=%d\n", mode);
+            break;
+    }
+    return scan_mode;
+}
+
 static int tvin_started = 0;
 static void stop_clone(void)
 {
 #ifdef CONFIG_TVIN_VIUIN
     if(tvin_started){
-        stop_tvin_service(0);
+        vdin_v4l2_ops_t * vops = get_vdin_v4l2_ops();
+        vops->stop_tvin_service(0);
         tvin_started=0;
     }
 #endif
@@ -2461,17 +2530,11 @@ static int start_clone(void)
 {
     int ret = -1;
 #ifdef CONFIG_TVIN_VIUIN
-    tvin_parm_t para;
+    vdin_parm_t para;
     const vinfo_t *info = get_current_vinfo();
+    vdin_v4l2_ops_t * vops = get_vdin_v4l2_ops();
     if(tvin_started){
-      
-     if(debug&0x200){
-     	vinfo_t *info2 = get_current_vinfo2();  
-      	if(strncmp(info2->name, "null", 4)==0){
-        	return 0; // dual display debug
-      	}
-	 }
-      stop_tvin_service(0);
+      vops->stop_tvin_service(0);
       tvin_started=0;
       ret = 0;
     }
@@ -2486,15 +2549,48 @@ static int start_clone(void)
             aml_set_reg32_bits(P_VPU_VIU_VENC_MUX_CTRL, 2, 8, 4); //reg0x271a,Enable VIU of ENC_P domain to VDIN;
         }
         clone_vpts_remainder = 0;
-        memset(&para,0,sizeof(tvin_parm_t));
-        para.fmt_info.h_active = info->width;
-        para.fmt_info.v_active = info->height;
-        para.port  = TVIN_PORT_VIU_ENCT;
-        para.fmt_info.fmt = info->mode; //TVIN_SIG_FMT_MAX+1;//TVIN_SIG_FMT_MAX+1;TVIN_SIG_FMT_CAMERA_1280X720P_30Hz
-        para.fmt_info.frame_rate = clone_frame_rate*10;
-        para.fmt_info.hsync_phase = 1;
-        para.fmt_info.vsync_phase  = 0;
-        start_tvin_service(0,&para);
+        memset(&para,0,sizeof(vdin_parm_t));
+        para.h_active = info->width;
+        para.v_active = info->height;
+        para.port	= TVIN_PORT_VIU;
+        para.fmt = TVIN_SIG_FMT_MAX; //TVIN_SIG_FMT_MAX+1;//TVIN_SIG_FMT_MAX+1;TVIN_SIG_FMT_CAMERA_1280X720P_30Hz
+        para.frame_rate = clone_frame_rate*10;
+        para.hsync_phase = 1;
+        para.vsync_phase  = 0;
+        para.hs_bp = 0;
+        para.vs_bp = 2;
+        para.cfmt = TVIN_YUV422;
+        para.scan_mode = vmode2scan_mode(info->mode);
+        if(TVIN_SCAN_MODE_INTERLACED == para.scan_mode){
+            para.v_active = para.v_active/2;
+        }
+        para.dfmt = TVIN_NV21;
+        if((vinfo) && (vinfo->width != 0) &&  (vinfo->height != 0)){
+            int dst_w = vinfo->width ;
+            int dst_h = vinfo->height;
+            if(vinfo->width<vinfo->height){
+                if((vinfo->width<=768)&&(vinfo->height<=1024)){
+                    dst_w = vinfo->width;
+                    dst_h = vinfo->height;
+                }else{
+                    dst_w = vinfo->height;
+                    dst_h = vinfo->width;
+                }
+                output_axis_adjust(vinfo->height,vinfo->width, (int *)&dst_h,(int *)&dst_w,0);
+            }else{
+                if((vinfo->height<=768)&&(vinfo->width<=1024)){
+                    dst_w = vinfo->width;
+                    dst_h = vinfo->height;
+                }
+                output_axis_adjust(vinfo->width,vinfo->height, (int *)&dst_w,(int *)&dst_h,0); 
+            }
+            para.dest_hactive = dst_w;
+            para.dest_vactive = dst_h;
+            if(TVIN_SCAN_MODE_INTERLACED == para.scan_mode){
+                para.dest_vactive = para.dest_vactive/2;
+            }
+        }
+        vops->start_tvin_service(0,&para);
         tvin_started = 1;
         printk("%s: source %dx%d\n", __func__,info->width, info->height);
         ret = 0;
@@ -2650,7 +2746,7 @@ static ssize_t frame_rate_show(struct class *cla, struct class_attribute* attr,
     u32 time = jiffies;
     u32 tmp = time;
     u32 rate = 0;
-    size_t ret;
+    size_t ret = 0;
     time -= last_frame_time;
     last_frame_time = tmp;
     last_frame_count = frame_count;
diff --git a/drivers/amlogic/display/vout2/nulldisp.c b/drivers/amlogic/display/vout2/nulldisp.c
index bee3b24b9b12..08d856b6c46b 100755
--- a/drivers/amlogic/display/vout2/nulldisp.c
+++ b/drivers/amlogic/display/vout2/nulldisp.c
@@ -74,12 +74,13 @@ static const vinfo_t *get_valid_vinfo(char  *mode)
 static vmode_t nulldisp_validate_vmode(char *mode)
 {
     const vinfo_t *info = get_valid_vinfo(mode);
+#ifndef CONFIG_AM_VIDEO2
     int viu1_select = aml_read_reg32(P_VPU_VIU_VENC_MUX_CTRL)&0x3;
-    
+#endif
 	DisableVideoLayer();
-	
+#ifndef CONFIG_AM_VIDEO2
     aml_set_reg32_bits (P_VPU_VIU_VENC_MUX_CTRL, (viu1_select+1)&0x3, 2, 2); //viu2_select should be different from viu1_select (to fix viu1 video smooth problem)
-
+#endif
     if (info)
         return info->mode;
 
diff --git a/drivers/amlogic/display/vout2/tvconf2.c b/drivers/amlogic/display/vout2/tvconf2.c
index a711b9888423..208178063455 100755
--- a/drivers/amlogic/display/vout2/tvconf2.c
+++ b/drivers/amlogic/display/vout2/tvconf2.c
@@ -70,8 +70,9 @@ SET_TV2_CLASS_ATTR(vdac_setting,parse_vdac_setting)
 
 static const tvmode_t vmode_tvmode_tab[] =
 {
-	TVMODE_480I, TVMODE_480CVBS,TVMODE_480P, TVMODE_576I,TVMODE_576CVBS, TVMODE_576P, TVMODE_720P, TVMODE_1080I, TVMODE_1080P,
-    TVMODE_720P_50HZ, TVMODE_1080I_50HZ, TVMODE_1080P_50HZ
+    TVMODE_480I, TVMODE_480I_RPT, TVMODE_480CVBS, TVMODE_480P, TVMODE_480P_RPT, TVMODE_576I, TVMODE_576I_RPT, TVMODE_576CVBS, TVMODE_576P, TVMODE_576P_RPT, TVMODE_720P, TVMODE_1080I, TVMODE_1080P,
+    TVMODE_720P_50HZ, TVMODE_1080I_50HZ, TVMODE_1080P_50HZ,TVMODE_1080P_24HZ, TVMODE_4K2K_30HZ, TVMODE_4K2K_25HZ, TVMODE_4K2K_24HZ, TVMODE_4K2K_SMPTE, 
+    TVMODE_VGA, TVMODE_SVGA, TVMODE_XGA, TVMODE_SXGA
 };
 
 
@@ -85,9 +86,22 @@ static const vinfo_t tv_info[] =
         .field_height      = 240,
         .aspect_ratio_num  = 4,
         .aspect_ratio_den  = 3,
-        .sync_duration_num = 5994,
-        .sync_duration_den = 100,
+        .sync_duration_num = 60,
+        .sync_duration_den = 1,
+        .video_clk         = 27000000,
     },
+    { /* VMODE_480I_RPT */
+        .name              = "480i_rpt",
+        .mode              = VMODE_480I_RPT,
+        .width             = 720,
+        .height            = 480,
+        .field_height      = 240,
+        .aspect_ratio_num  = 4,
+        .aspect_ratio_den  = 3,
+        .sync_duration_num = 60,
+        .sync_duration_den = 1,
+        .video_clk         = 27000000,
+     },
      { /* VMODE_480CVBS*/
         .name              = "480cvbs",
         .mode              = VMODE_480CVBS,
@@ -96,8 +110,9 @@ static const vinfo_t tv_info[] =
         .field_height      = 240,
         .aspect_ratio_num  = 4,
         .aspect_ratio_den  = 3,
-        .sync_duration_num = 5994,
-        .sync_duration_den = 100,
+        .sync_duration_num = 60,
+        .sync_duration_den = 1,
+        .video_clk         = 27000000,
     },
     { /* VMODE_480P */
         .name              = "480p",
@@ -107,8 +122,21 @@ static const vinfo_t tv_info[] =
         .field_height      = 480,
         .aspect_ratio_num  = 4,
         .aspect_ratio_den  = 3,
-        .sync_duration_num = 5994,
-        .sync_duration_den = 100,
+        .sync_duration_num = 60,
+        .sync_duration_den = 1,
+        .video_clk         = 27000000,
+    },
+    { /* VMODE_480P_RPT */
+        .name              = "480p_rpt",
+        .mode              = VMODE_480P_RPT,
+        .width             = 720,
+        .height            = 480,
+        .field_height      = 480,
+        .aspect_ratio_num  = 4,
+        .aspect_ratio_den  = 3,
+        .sync_duration_num = 60,
+        .sync_duration_den = 1,
+        .video_clk         = 27000000,
     },
     { /* VMODE_576I */
         .name              = "576i",
@@ -120,6 +148,19 @@ static const vinfo_t tv_info[] =
         .aspect_ratio_den  = 3,
         .sync_duration_num = 50,
         .sync_duration_den = 1,
+        .video_clk         = 27000000,
+    },
+    { /* VMODE_576I_RPT */
+        .name              = "576i_rpt",
+        .mode              = VMODE_576I_RPT,
+        .width             = 720,
+        .height            = 576,
+        .field_height      = 288,
+        .aspect_ratio_num  = 4,
+        .aspect_ratio_den  = 3,
+        .sync_duration_num = 50,
+        .sync_duration_den = 1,
+        .video_clk         = 27000000,
     },
     { /* VMODE_576I */
         .name              = "576cvbs",
@@ -131,6 +172,7 @@ static const vinfo_t tv_info[] =
         .aspect_ratio_den  = 3,
         .sync_duration_num = 50,
         .sync_duration_den = 1,
+        .video_clk         = 27000000,
     },
     { /* VMODE_576P */
         .name              = "576p",
@@ -142,6 +184,19 @@ static const vinfo_t tv_info[] =
         .aspect_ratio_den  = 3,
         .sync_duration_num = 50,
         .sync_duration_den = 1,
+        .video_clk         = 27000000,
+    },
+    { /* VMODE_576P_RPT */
+        .name              = "576p_rpt",
+        .mode              = VMODE_576P_RPT,
+        .width             = 720,
+        .height            = 576,
+        .field_height      = 576,
+        .aspect_ratio_num  = 4,
+        .aspect_ratio_den  = 3,
+        .sync_duration_num = 50,
+        .sync_duration_den = 1,
+        .video_clk         = 27000000,
     },
     { /* VMODE_720P */
         .name              = "720p",
@@ -153,6 +208,7 @@ static const vinfo_t tv_info[] =
         .aspect_ratio_den  = 9,
         .sync_duration_num = 60,
         .sync_duration_den = 1,
+        .video_clk         = 74250000,
     },
     { /* VMODE_1080I */
         .name              = "1080i",
@@ -164,6 +220,7 @@ static const vinfo_t tv_info[] =
         .aspect_ratio_den  = 9,
         .sync_duration_num = 60,
         .sync_duration_den = 1,
+        .video_clk         = 74250000,
     },
     { /* VMODE_1080P */
         .name              = "1080p",
@@ -175,6 +232,7 @@ static const vinfo_t tv_info[] =
         .aspect_ratio_den  = 9,
         .sync_duration_num = 60,
         .sync_duration_den = 1,
+        .video_clk         = 148500000,
     },
     { /* VMODE_720P_50hz */
         .name              = "720p50hz",
@@ -186,6 +244,7 @@ static const vinfo_t tv_info[] =
         .aspect_ratio_den  = 9,
         .sync_duration_num = 50,
         .sync_duration_den = 1,
+        .video_clk         = 74250000,
     },
     { /* VMODE_1080I_50HZ */
         .name              = "1080i50hz",
@@ -197,6 +256,7 @@ static const vinfo_t tv_info[] =
         .aspect_ratio_den  = 9,
         .sync_duration_num = 50,
         .sync_duration_den = 1,
+        .video_clk         = 74250000,
     },
     { /* VMODE_1080P_50HZ */
         .name              = "1080p50hz",
@@ -208,8 +268,140 @@ static const vinfo_t tv_info[] =
         .aspect_ratio_den  = 9,
         .sync_duration_num = 50,
         .sync_duration_den = 1,
+        .video_clk         = 148500000,
+    },
+    { /* VMODE_1080P_24HZ */
+        .name              = "1080p24hz",
+        .mode              = VMODE_1080P_24HZ,
+        .width             = 1920,
+        .height            = 1080,
+        .field_height      = 1080,
+        .aspect_ratio_num  = 16,
+        .aspect_ratio_den  = 9,
+        .sync_duration_num = 24,
+        .sync_duration_den = 1,
+        .video_clk         = 74250000,
+    },
+    { /* VMODE_4K2K_30HZ */
+        .name              = "4k2k30hz",
+        .mode              = TVMODE_4K2K_30HZ,
+        .width             = 3840,
+        .height            = 2160,
+        .field_height      = 2160,
+        .aspect_ratio_num  = 16,
+        .aspect_ratio_den  = 9,
+        .sync_duration_num = 30,
+        .sync_duration_den = 1,
+        .video_clk         = 297000000,
+    },
+    { /* VMODE_4K2K_25HZ */
+        .name              = "4k2k25hz",
+        .mode              = TVMODE_4K2K_25HZ,
+        .width             = 3840,
+        .height            = 2160,
+        .field_height      = 2160,
+        .aspect_ratio_num  = 16,
+        .aspect_ratio_den  = 9,
+        .sync_duration_num = 25,
+        .sync_duration_den = 1,
+        .video_clk         = 297000000,
+    },
+    { /* VMODE_4K2K_24HZ */
+        .name              = "4k2k24hz",
+        .mode              = TVMODE_4K2K_24HZ,
+        .width             = 3840,
+        .height            = 2160,
+        .field_height      = 2160,
+        .aspect_ratio_num  = 16,
+        .aspect_ratio_den  = 9,
+        .sync_duration_num = 24,
+        .sync_duration_den = 1,
+        .video_clk         = 297000000,
+    },
+    { /* VMODE_4K2K_SMPTE */
+        .name              = "4k2ksmpte",
+        .mode              = TVMODE_4K2K_SMPTE,
+        .width             = 4096,
+        .height            = 2160,
+        .field_height      = 2160,
+        .aspect_ratio_num  = 16,
+        .aspect_ratio_den  = 9,
+        .sync_duration_num = 24,
+        .sync_duration_den = 1,
+        .video_clk         = 297000000,
+    },
+    { /* VMODE_vga */
+        .name              = "vga",
+        .mode              = VMODE_VGA,
+        .width             = 640,
+        .height            = 480,
+        .field_height      = 240,
+        .aspect_ratio_num  = 4,
+        .aspect_ratio_den  = 3,
+        .sync_duration_num = 60,
+        .sync_duration_den = 1,
+        .video_clk         = 25175000,
+    },
+    { /* VMODE_SVGA */
+        .name              = "svga",
+        .mode              = VMODE_SVGA,
+        .width             = 800,
+        .height            = 600,
+        .field_height      = 600,
+        .aspect_ratio_num  = 4,
+        .aspect_ratio_den  = 3,
+        .sync_duration_num = 60,
+        .sync_duration_den = 1,
+        .video_clk         = 40000000,
+    },
+    { /* VMODE_XGA */
+        .name              = "xga",
+        .mode              = VMODE_XGA,
+        .width             = 1024,
+        .height            = 768,
+        .field_height      = 768,
+        .aspect_ratio_num  = 4,
+        .aspect_ratio_den  = 3,
+        .sync_duration_num = 60,
+        .sync_duration_den = 1,
+        .video_clk         = 65000000,
+    },
+    { /* VMODE_sxga */
+        .name              = "sxga",
+        .mode              = VMODE_SXGA,
+        .width             = 1280,
+        .height            = 1024,
+        .field_height      = 1024,
+        .aspect_ratio_num  = 5,
+        .aspect_ratio_den  = 4,
+        .sync_duration_num = 60,
+        .sync_duration_den = 1,
+        .video_clk         = 108000000,
+    },
+	{ /* VMODE_wsxga */
+        .name              = "wsxga",
+        .mode              = VMODE_WSXGA,
+        .width             = 1440,
+        .height            = 900,
+        .field_height      = 900,
+        .aspect_ratio_num  = 8,
+        .aspect_ratio_den  = 5,
+        .sync_duration_num = 60,
+        .sync_duration_den = 1,
+        .video_clk         = 88750000,
+    },
+	{ /* VMODE_fhdvga */
+        .name              = "fhdvga",
+        .mode              = VMODE_FHDVGA,
+        .width             = 1920,
+        .height            = 1080,
+        .field_height      = 1080,
+        .aspect_ratio_num  = 16,
+        .aspect_ratio_den  = 9,
+        .sync_duration_num = 60,
+        .sync_duration_den = 1,
+        .video_clk         = 148500000,
     },
-
 };
 
 static const struct file_operations am_tv_fops = {
@@ -247,7 +439,7 @@ static const vinfo_t *tv_get_current_info(void)
 
 static int tv_set_current_vmode(vmode_t mod)
 {
-	if ((mod&VMODE_MODE_BIT_MASK)> VMODE_1080P_50HZ)
+	if ((mod&VMODE_MODE_BIT_MASK)> VMODE_SXGA)
 		return -EINVAL;
 
 	info->vinfo = &tv_info[mod];
diff --git a/drivers/amlogic/display/vout2/tvmode.h b/drivers/amlogic/display/vout2/tvmode.h
index 28902dd09d30..a39a0a7c30af 100755
--- a/drivers/amlogic/display/vout2/tvmode.h
+++ b/drivers/amlogic/display/vout2/tvmode.h
@@ -27,17 +27,30 @@
 
 typedef enum {
     TVMODE_480I  = 0,
+    TVMODE_480I_RPT,
     TVMODE_480CVBS,
     TVMODE_480P  ,
+    TVMODE_480P_RPT,
     TVMODE_576I  ,
+    TVMODE_576I_RPT,
     TVMODE_576CVBS,
     TVMODE_576P  ,
+    TVMODE_576P_RPT,
     TVMODE_720P  ,
     TVMODE_1080I ,
     TVMODE_1080P ,
     TVMODE_720P_50HZ ,
     TVMODE_1080I_50HZ ,
     TVMODE_1080P_50HZ ,
+    TVMODE_1080P_24HZ ,
+    TVMODE_4K2K_30HZ ,
+    TVMODE_4K2K_25HZ ,
+    TVMODE_4K2K_24HZ ,
+    TVMODE_4K2K_SMPTE ,
+    TVMODE_VGA ,
+    TVMODE_SVGA,
+    TVMODE_XGA,
+    TVMODE_SXGA,
     TVMODE_MAX   
 } tvmode_t;
 
diff --git a/drivers/amlogic/display/vout2/tvoutc2.c b/drivers/amlogic/display/vout2/tvoutc2.c
index e740d62217dd..cb73893248cc 100755
--- a/drivers/amlogic/display/vout2/tvoutc2.c
+++ b/drivers/amlogic/display/vout2/tvoutc2.c
@@ -199,9 +199,12 @@ int tvoutc_setclk2(tvmode_t mode)
 	switch(mode)
 	{
 		case TVMODE_480I:
+		case TVMODE_480I_RPT:
 		case TVMODE_480CVBS:
 		case TVMODE_480P:
+		case TVMODE_480P_RPT:
 		case TVMODE_576I:
+		case TVMODE_576I_RPT:
 		case TVMODE_576CVBS:
 		case TVMODE_576P:
 			  setreg(&sd[xtal]);
@@ -250,7 +253,32 @@ int tvoutc_setmode2(tvmode_t mode)
 		case TVMODE_576I:
 		case TVMODE_576CVBS:
 		case TVMODE_576P:
-        //WRITE_MPEG_REG(HHI_VIID_CLK_DIV, (READ_MPEG_REG(HHI_VIID_CLK_DIV)&(~(0xff)))|0x3); // reg 0x104a
+			  //WRITE_MPEG_REG(HHI_VIID_CLK_DIV, (READ_MPEG_REG(HHI_VIID_CLK_DIV)&(~(0xff)))|0x3); // reg 0x104a 
+			  WRITE_VCBUS_REG_BITS(VPU_VIU_VENC_MUX_CTRL, 1, 2, 2); //reg0x271a, select ENCT to VIU2
+			  WRITE_CBUS_REG_BITS(HHI_VIID_DIVIDER_CNTL, 0x03, 0, 2);
+			  WRITE_CBUS_REG_BITS(HHI_VID2_PLL_CNTL5, 1, 16, 1);
+			  aml_write_reg32(P_HHI_VID2_PLL_CNTL, 0x41000032);//--set vid2 pll < 1.7G
+			  aml_write_reg32(P_HHI_VID2_PLL_CNTL2, 0x0421a000);
+			  aml_write_reg32(P_HHI_VID2_PLL_CNTL3, 0xca45b823);
+			  aml_write_reg32(P_HHI_VID2_PLL_CNTL4, 0xd4000d67);
+			  aml_write_reg32(P_HHI_VID2_PLL_CNTL5, 0x01700001);//--
+			  WRITE_CBUS_REG_BITS(HHI_VID2_PLL_CNTL, 0x01, 24, 5);//clock tree N=1
+			  WRITE_CBUS_REG_BITS(HHI_VID2_PLL_CNTL, 0x36, 0, 9);//clock tree M=54
+			  WRITE_CBUS_REG_BITS(HHI_VID2_PLL_CNTL, 0x00, 9, 2);//clock tree OD=0
+			  WRITE_CBUS_REG_BITS(HHI_VID2_PLL_CNTL5, 0x01, 24, 1);//select CLK1 = 24x54 = 1296M
+			  WRITE_CBUS_REG_BITS(HHI_DSI_LVDS_EDP_CNTL1, 0x00, 4, 1);
+			  WRITE_CBUS_REG_BITS(HHI_VIID_DIVIDER_CNTL, 0x03, 15, 2);
+			  WRITE_CBUS_REG_BITS(HHI_VIID_DIVIDER_CNTL, 0x05, 4, 3);//select 6
+			  WRITE_CBUS_REG_BITS(HHI_VIID_DIVIDER_CNTL, 0x00, 8, 2);//select 0
+			  WRITE_CBUS_REG_BITS(HHI_VIID_DIVIDER_CNTL, 0x00, 12, 3);//select 0 clock = 1296/6 = 216M
+			  WRITE_CBUS_REG_BITS(HHI_VIID_CLK_CNTL, 0x0c, 16, 4);//select vid2_pll_clk & open bit19
+			  WRITE_CBUS_REG_BITS(HHI_VIID_CLK_DIV, 0x03, 0, 4);//select 4 clock = 216/4 = 54M
+			  WRITE_CBUS_REG_BITS(HHI_VIID_CLK_CNTL, 0x07, 0, 5);//open div1 div2 div4
+			  WRITE_CBUS_REG_BITS(HHI_VID_CLK_DIV, 0x09, 28, 4);//reg 0x1059, select v2_clk_div2 for cts_enci_clk , 27MHz
+			  WRITE_CBUS_REG_BITS(HHI_VIID_CLK_DIV, 0x09, 28, 4); //0x104a, select v2_clk_div2 for cts_vdac_clk[0] 
+			  WRITE_CBUS_REG_BITS(HHI_VIID_CLK_DIV, 0x09, 24, 4); //0x104a, select v2_clk_div2 for cts_vdac_clk[1] (DAC3)
+			  WRITE_CBUS_REG_BITS(HHI_VID2_PLL_CNTL, 0x01, 29, 1);//reset
+			  WRITE_CBUS_REG_BITS(HHI_VID2_PLL_CNTL, 0x00, 29, 1);//clean reset
 			  break;
 		case TVMODE_720P:
 		case TVMODE_720P_50HZ:
@@ -265,12 +293,13 @@ int tvoutc_setmode2(tvmode_t mode)
 		default:
 			printk(KERN_ERR "unsupport tv mode,video clk is not set!!\n");	
 	}
-    
-        //WRITE_MPEG_REG(HHI_VIID_CLK_CNTL, 0x8001f); //reg 0x104b, select vid_pll_clk as source of v2_clk_divN
-        //WRITE_MPEG_REG(HHI_VID_CLK_DIV, (READ_MPEG_REG(HHI_VID_CLK_DIV)&(~(0xff<<24)))|(0x88<<24)); // reg 0x1059, select cts_encp_clk and cts_enci_clk from v2_clk_div1
-        aml_set_reg32_bits(P_VPU_VIU_VENC_MUX_CTRL, 2, 2, 2); //reg0x271a, select ENCP to VIU2
-
-    
+    //WRITE_MPEG_REG(HHI_VIID_CLK_CNTL, 0x8001f); //reg 0x104b, select vid_pll_clk as source of v2_clk_divN
+    //WRITE_MPEG_REG(HHI_VID_CLK_DIV, (READ_MPEG_REG(HHI_VID_CLK_DIV)&(~(0xff<<24)))|(0x88<<24)); // reg 0x1059, select cts_encp_clk and cts_enci_clk from v2_clk_div1
+    //aml_set_reg32_bits(P_VPU_VIU_VENC_MUX_CTRL, 2, 2, 2); //reg0x271a, select ENCP to VIU2
+    aml_write_reg32(P_HHI_VDAC_CNTL0, 0x00000001);
+    aml_write_reg32(P_HHI_VDAC_CNTL1, 0x00000000);
+    aml_write_reg32(P_ENCI_MACV_N0, 0x00000000);
+    aml_write_reg32(P_HHI_GCLK_OTHER, aml_read_reg32(P_HHI_GCLK_OTHER)|(1<<8));
     aml_write_reg32(P_VPP2_POSTBLEND_H_SIZE, tvinfoTab[mode].xres);
     
 // For debug only
diff --git a/drivers/amlogic/display/vout2/tvregs.h b/drivers/amlogic/display/vout2/tvregs.h
index 21bf1e73d225..56da502ab4f2 100755
--- a/drivers/amlogic/display/vout2/tvregs.h
+++ b/drivers/amlogic/display/vout2/tvregs.h
@@ -366,7 +366,7 @@ static const reg_t tvregs_480p[] = {
 };
 
 static const reg_t tvregs_576i[] = {
-    {VENC_VDAC_SETTING,          0xff,  },
+    {P_VENC_VDAC_SETTING,          0xff,  },
 //	{VCLK_SD},
     {P_ENCI_CFILT_CTRL,                 0x12,    },
     {P_ENCI_CFILT_CTRL2,                 0x12,    },
@@ -417,7 +417,7 @@ static const reg_t tvregs_576i[] = {
 };
 
 static const reg_t tvregs_576cvbs[] = {
-    {VENC_VDAC_SETTING,          0xff,  },
+    {P_VENC_VDAC_SETTING,          0xff,  },
 //	{VCLK_SD},
     {P_ENCI_CFILT_CTRL,                 0x12,    },
     {P_ENCI_CFILT_CTRL2,                 0x12,    },
@@ -732,36 +732,83 @@ static const reg_t tvregs_1080p_50hz[] = {
     {MREG_END_MARKER,            0      }
 };
 
+static const reg_t tvregs_1080p_24hz[] = {
+};
+
+static const reg_t tvregs_4k2k_30hz[] = {
+};
+
+static const reg_t tvregs_4k2k_25hz[] = {
+};
+
+static const reg_t tvregs_4k2k_24hz[] = {
+};
+
+static const reg_t tvregs_4k2k_smpte[] = {
+};
+
+static const reg_t tvregs_vga_640x480[] = {
+};
+
+static const reg_t tvregs_svga_800x600[]={
+};
+
+static const reg_t tvregs_xga_1024x768[] = {
+};
 
 /* The sequence of register tables items must match the enum define in tvmode.h */
 static const reg_t *tvregsTab[] = {
+    tvregs_480i,
     tvregs_480i,
     tvregs_480cvbs,		
     tvregs_480p,
+    tvregs_480p,
+    tvregs_576i,
     tvregs_576i,
     tvregs_576cvbs,
     tvregs_576p,
+    tvregs_576p,
     tvregs_720p,
     tvregs_1080i,       //Adjust tvregs_* sequences and match the enum define in tvmode.h
     tvregs_1080p,
     tvregs_720p_50hz,
     tvregs_1080i_50hz,
-    tvregs_1080p_50hz
+    tvregs_1080p_50hz,
+    tvregs_1080p_24hz,
+    tvregs_4k2k_30hz,
+    tvregs_4k2k_25hz,
+    tvregs_4k2k_24hz,
+    tvregs_4k2k_smpte,
+    tvregs_vga_640x480,
+    tvregs_svga_800x600,
+    tvregs_xga_1024x768
 };
 
 static const tvinfo_t tvinfoTab[] = {
     {.xres =  720, .yres =  480, .id = "480i"},
+    {.xres =  720, .yres =  480, .id = "480i_rpt"},
     {.xres =  720, .yres =  480, .id = "480cvbs"},		
     {.xres =  720, .yres =  480, .id = "480p"},
+    {.xres =  720, .yres =  480, .id = "480p_rpt"},
     {.xres =  720, .yres =  576, .id = "576i"},
+    {.xres =  720, .yres =  576, .id = "576i_rpt"},
     {.xres =  720, .yres =  576, .id = "576cvbs"},
     {.xres =  720, .yres =  576, .id = "576p"},
+    {.xres =  720, .yres =  576, .id = "576p_prt"},
     {.xres = 1280, .yres =  720, .id = "720p"},
     {.xres = 1920, .yres = 1080, .id = "1080i"},
     {.xres = 1920, .yres = 1080, .id = "1080p"},
     {.xres = 1280, .yres =  720, .id = "720p50hz"},
     {.xres = 1920, .yres = 1080, .id = "1080i50hz"},
-    {.xres = 1920, .yres = 1080, .id = "1080p50hz"}
+    {.xres = 1920, .yres = 1080, .id = "1080p50hz"},
+    {.xres = 1920, .yres = 1080, .id = "1080p24hz"},
+    {.xres = 3840, .yres = 2160, .id = "4k2k30hz"},
+    {.xres = 3840, .yres = 2160, .id = "4k2k25hz"},
+    {.xres = 3840, .yres = 2160, .id = "4k2k24hz"},
+    {.xres = 4096, .yres = 2160, .id = "4k2ksmpte"},
+    {.xres = 640, .yres = 480, .id = "vga"},
+    {.xres = 800, .yres = 600, .id = "svga"},
+    {.xres = 1024, .yres = 768, .id = "xga"},
 };
 
 static inline void setreg(const reg_t *r)
diff --git a/drivers/amlogic/display/vout2/vout2_serve.c b/drivers/amlogic/display/vout2/vout2_serve.c
index 13c712381366..82aa9fa399bb 100755
--- a/drivers/amlogic/display/vout2/vout2_serve.c
+++ b/drivers/amlogic/display/vout2/vout2_serve.c
@@ -67,9 +67,31 @@ static int s_venc_mux = 0;
 **	sysfs impletement part  
 **
 ******************************************************************/
-static  void   func_default_null(char  *str)
+static  int   func_default_null(char  *str)
 {
-	return ;
+#if MESON_CPU_TYPE == MESON_CPU_TYPE_MESON8
+    int val;
+    unsigned int cntl0=0, cntl1=0;
+    int r = sscanf(str, "%d", &val);
+    if (r != 1) {
+        return -EINVAL;
+    }
+    if ((val < 0) || (val > 1)) {
+        return -EINVAL;
+    }
+    if(!val){
+        cntl0 = 0;
+        cntl1 = 8;
+        WRITE_MPEG_REG(HHI_VDAC_CNTL0, cntl0);//close cvbs
+        WRITE_MPEG_REG(HHI_VDAC_CNTL1, cntl1);
+    }else{
+        cntl0 = 1;
+        cntl1 = 0;
+        WRITE_MPEG_REG(HHI_VDAC_CNTL0, cntl0);//open cvbs
+        WRITE_MPEG_REG(HHI_VDAC_CNTL1, cntl1);
+    }
+#endif
+    return 0;
 }
 static   int* parse_para(char *para,char   *para_num)
 {
diff --git a/drivers/amlogic/display/vout2/vout_serve.h b/drivers/amlogic/display/vout2/vout_serve.h
index bd0aac5afade..5ad79a33f900 100755
--- a/drivers/amlogic/display/vout2/vout_serve.h
+++ b/drivers/amlogic/display/vout2/vout_serve.h
@@ -77,7 +77,7 @@ static  void  write_reg(char *para);
 #endif
 static  void  set_vout_mode(char *mode) ;
 static void  set_vout_window(char *para) ;
-static  void   func_default_null(char  *str);
+static  int   func_default_null(char  *str);
 
 
 #endif
diff --git a/drivers/amlogic/vfm/vfm.c b/drivers/amlogic/vfm/vfm.c
index e2bfd7b78b71..82510877cacd 100755
--- a/drivers/amlogic/vfm/vfm.c
+++ b/drivers/amlogic/vfm/vfm.c
@@ -262,7 +262,7 @@ static void vfm_init(void)
 #endif
 #ifdef CONFIG_TVIN_VIUIN
     char def_ext_id[] = "default_ext";
-    char def_ext_name_chain[] = "vdin amvideo2";
+    char def_ext_name_chain[] = "vdin0 amvideo2";
 #else
 #ifdef CONFIG_AMLOGIC_VIDEOIN_MANAGER
     char def_ext_id[] = "default_ext";
-- 
2.19.0

